doc/config.md

## config.rb の説明

* config ファイルは、次の優先度で読み込まれます。

  * 環境変数 RASPIREC_CONF で指定したファイル(拡張子は .rb)
  * $HOME/.config/raspirec/config.rb


* 配列の定義で、`%w( 1 2 3 )` は `[ "1", "2", "3" ]` と等価です。

## 各パラメータの意味は次の通りです。

#### httpd

| 定数名             | 説明 |
|--------------------|------|
|Http_port|httpd のポート番号を指定する。<BR>WEBインターフェースには http://XXX.YYY.ZZZ:${Http_port}/ でアクセスする。|

#### ディレクトリ、ファイル関係

| 定数名             | 説明 |
|--------------------|------|
|Recpt1_cmd|recpt1 コマンドの path を指定する。|
|Recpt1_opt|recpt1 コマンドのオプションを指定する。b25 デコードをする場合は --b25 を指定する。|
|Epgdump|epgdump コマンドの path を指定する。|
|BaseDir|raspirec がインストールされているディレクトリを設定する。( raspirec.rb があるディレクトリ )|
|DataDir|データベースや録画したファイルの置き場所を指定する。|

#### 録画タイミング関係

| 定数名             | 説明 |
|--------------------|------|
|Start_margin|番組開始前の録画マージンを指定(秒)。録画は PC の時計を基準に、開始します。|
|After_margin|番組終了後の録画マージンを指定(秒)。録画は PC の時計を基準に、終了します。|
|Gap_time|録画が連続し、前番組が時短になった場合の、前番組録画終了から次番組開始間隔を指定(秒)。<br>前番組は Start_margin +  Gap_time の秒数だけ録画時間が削られます。<br>TVチューナー依存ですが、あまり短いと動作が不安定になります。|

#### チューナー関係

| 定数名             | 説明 |
|--------------------|------|
|GR_tuner_num|地デジチュナー数|
|BSCS_tuner_num|BSCSチュナー数|
|GBC_tuner_num|地デジ/BS/CS チューナー数|
|Total_tuner_limit|トータルチュナー数制限を掛ける場合に指定する。使用しない場合は false|

#### EPG関係

| 定数名             | 説明 |
|--------------------|------|
|GR_EPG_channel|地デジ EPG 受信局を設定する。<br>東京スカイツリーならば%w( 27 26 25 24 22 23 21 16 ) になります。<a href="https://www.maspro.co.jp/contact/channel.pdf#page=4"> 参考 </a>|
|BS_EPG_channel|BS EPG 受信局を設定する。BSを受信しない場合は空にする。<br>通常の番組情報は、どれか 1局だけ受信すれば十分だが、詳細情報を取得するには、その情報を取りたい局を指定する必要がある。|
|CS_EPG_channel|CS EPG 受信局を設定する。CSを受信しない場合は空にする。|
|GR_EpgRsvTime|地デジ EPG受信時間 (秒)|
|BS_EpgRsvTime|BS EPG受信時間 (秒)|
|CS_EpgRsvTime|CS EPG受信時間 (秒)|
|EPGperiod|EPG 取得周期 (H)|
|EpgBanTime|EPG の取得禁止時間帯の指定。(24H制 時間単位) 1時から5時までを禁止したい場合は、[ 1,2,3,4,5 ] と指定する。使用しない場合は nil|
|EPG_tuner_limit|EPG取得時の使用チュナー数に制限を掛ける場合に指定する。使用しない場合は false|


#### 自動録画延長機能
| 定数名             | 説明 |
|--------------------|------|
|AutoRecExt          | 自動録画延長機能  true で有効。 詳細は <a href="https://kaikoma-soft.github.io/raspirec-RecExt.html"> こちら </a> を参照 |
|ARE_sampling_time   | 番組終了の n 秒前に EPG 採取 |
|ARE_epgdump_opt     | 最後尾切り出しの epgdump のオプション |


#### ダイアログのオプション初期値

| 定数名             | 説明 |
|--------------------|------|
|D_FreeOnly|無料放送のみ|
|D_dedupe|重複予約は無効化する|
|D_jitan|チューナー不足の場合に時短を許可|

#### TSファイル転送

| 定数名             | 説明 |
|--------------------|------|
|TSFT|true=転送機能有効 true以外=無効<br>この機能を使うには、送り先のホストに対して,パスワードなしでssh,scpアクセス可能なように設定されていることが必要です。|
|TSFT_host|送り先 ホスト名|
|TSFT_user|送り先 login名|
|TSFT_toDir|送り先Dir|
|TSFT_rate|想定転送速度 ( Mbyte/秒 )<br> この数字を使って、空き時間に転送する／しないを判断します。|

#### HLS モニタ機能

| 定数名             | 説明 |
|--------------------|------|
|MonitorFunc|true=モニタ機能を有効, false=無効。 無効の場合は、下記のパラメータは無視される。設定方法は <a href="https://kaikoma-soft.github.io/raspirec-option.html"> こちら </a> を参照して下さい。|
|StreamDir|ストリーム出力先ディレクトリ|
|MonitorWidth|モニタ画面の横幅|
|HlsConvCmd|HLS変換コマンド|

#### mpv モニタ機能

| 定数名             | 説明 |
|--------------------|------|
|MPMonitor|true = mpv モニタ機能を有効、false = 無効 <br>無効の場合は、下記のパラメータは無視される。設定方法は <a href="https://kaikoma-soft.github.io/raspirec-option.html"> こちら </a> を参照して下さい。|
|DevAutoDetection| DeviceList_GR, DeviceList_BSCS, DeviceList_GBC を自動設定する。true = 有効。|
|DeviceList_GR|地デジのデバイスファイルを指定する。|
|DeviceList_BSCS|BS,CS のデバイスファイルを指定する。|
|DeviceList_GBC | 地デジ,BS,CS のデバイスファイルを指定する。|
|Mpv_cmd|mpv の実行ファイル名(絶対パス)|
|Mpv_opt|mpv の引数を指定する。
|RemoteMonitor|表示するマシンが別の場合 ture, 同一の場合に false|
|UDPbasePort|使用する UDP のポート番号、デバイスの数だけプラスされる。|
|XServerName|RemoteMonitor が true の場合に、mpvを実行、表示するマシンのマシン名を設定する。|
|RecHostName|RemoteMonitor が true の場合に、チューナー(raspirecが実行されている)のマシン名を設定する。|
|Lsof_cmd|lsof コマンドへのパス。Ubuntu では /usr/bin/lsof |
|Browser_cmd| 番組表を表示するブラウザの実行ファイル名(絶対パス) |
|RaspirecTV_font|フォント指定|
|RaspirecTV_GEO|座標指定 ( WxH+X+Y or X+Y )|
|RaspirecTV_SOCAT|socat の実行ファイル名(絶対パス)(RemoteMonitor == true の時のみ)|

#### その他

| 定数名             | 説明 |
|--------------------|------|
|TSnameFormat|TSファイル名の生成ルール。詳細は 「コントロールパネル／補足説明」を参照して下さい。|
|LogSaveDay|ログの保持期間(日)|
|RsvHisSaveDay|録画済み記録の保持期間(日)|
|DiskKeepPercent|録画したTSファイルを古い順に削除して指定したDisk容量(%)を確保する。(指定するのは空き容量)<br>機能を無効にする場合は false を指定する。|
|Local_jquery|オフライン環境で動作させる為に jquery, materialize のライブラリをローカルにコピーした場合に、true にする。<br>詳細は doc/jquery_local.md を参照の事。 通常は false|
|StationPage|番組表で、1ページ当たりの放送局数 (個) の初期値|
|Debug|ture で Debug モード。ログファイルを出力するようになる。オプションで、 -d を指定するのと同じ。|
|Debug_mem|ture で メモリの消費量をモニタするようになる。|
|TitleRegex|「自動予約」ボタンを押して番組検索に遷移した時に、番組タイトルから「題名」を生成する為の、余計な文字を削除する正規表現の配列|
|SearchStringRegex|「自動予約」ボタンを押して番組検索に遷移した時に、番組タイトルから「検索文字列」を生成する為の、余計な文字を削除する正規表現の配列|
|EpgPatchEnable|EPGPatch機能を有効にする。 false = 無効。デフォルトは有効|

CHANGELOG.txt

ver 1.3.0         2021/09/23
  * 機能追加
    * 録画中の番組が延長になった場合に、自動で追従する機能を追加
      詳細は https://kaikoma-soft.github.io/src/raspirec-RecExt.html を参照
    * コントロールパネルに、ログファイルの強制ローテートを追加

  * 変更
    * ラズパイ3B+ で、高負荷時に SQLite3::Busy による Deadlock が発生
      するので、トランザクション関係を見直し。
    * 従来は録画中の Disk容量確保はしなかったが、するように。(DiskFull回避優先)
    
  * 注意
    上記機能追加に伴い、録画終了の方法を recpt1/recdvb 任せから、
    raspirec の時計で終了させる方法に変更した。
    その為、従来より After_margin を 2秒程度遅くしなければならない。

ver 1.2.0         2021/07/04
  * 機能追加
    * 「mpvモニタ機能」を独立した GUI にしたツール raspirecTV.rb を追加。
       詳細は https://kaikoma-soft.github.io/src/raspirec-TV.html を参照
     
  * 変更
    * config の定数 MPlayer_cmd を廃止し、Mpv_cmd と Mpv_opt を新設


ver 1.1.1         2021/05/30
  *機能追加
    * EPG取得時の使用チューナー数を制限する 定数 EPG_tuner_limit を
      config.rb に導入


ver 1.1.0         2021/05/27
  ・機能追加
    + PX-MLT8PE のような 3波(地デジ/BS/CS)チューナーへの対応
        (注：実物は所有していないので、ダミー環境でのテストのみ )
       * config.rb に GBC_tuner_num, DeviceList_GBC を追加
       * 予約スケジュール計算ルーチンの全面書き換え。
       * mpv モニタで、3波を選択可能に。
    + DeviceList_* にデバイスの自動検出機能の追加( DevAutoDetection )
      ( 詳細は、コントールパネル->補足説明->デバイスの自動検出 を参照 )
    + DB ファイルのバックアップを自動で残すように。(一週間分)
    
  ・不具合修正
    + https://github.com/kaikoma-soft/raspirec/issues/9 で指摘された
      番組表等のツールチップの不具合を Materialize のツールチップから
      自前のツールチップに変更して対策。
    + earth_pt1 + recdvb の環境で、mpv モニタが動作していなかった件を
      修正

  ・ 改良
    + 自動予約で一致件数が多い場合に、時間が掛かっていたのを改善。
    + EPG 取得時に全部のチューナーを使うように。
      (従来は地デジ、BS/CS で、1本ずつ)
    + 録画開始直前の EPG取得
       * 取得タイミングを録画開始 20分前から 10分前に変更
       * 取得時間を config で指定した値が 60秒以上なら 60秒に制限して取得。
       * config で指定した局では無く録画対象の局から直接取得する様に。


ver 1.0.4         2021/05/01
  ・ 環境により EpgPatch機能が不要になる事が判明したので、 EpgPatch
     機能を無効にするパラメータ EpgPatchEnable を config.rb に追加した。
     (デフォルトは有効)

     具体的な設定は、放送大学ex (BS11_2, 231ch) が 
     BS11_1 で受信できる場合は   EpgPatchEnable = true
     BS11_2 で受信できる場合は   EpgPatchEnable = false
     又は
     px4_drv + recpt1 の場合は   EpgPatchEnable = true
     pt1_drv + recpt1 の場合は   EpgPatchEnable = true
     earth_pt1 + recdvb の場合は EpgPatchEnable = false

ver 1.0.3         2021/04/13
  ・4/13 のトランスポンダの移動で、ＢＳ釣りビジョンに EPG とのズレが生
    じたので、EpgPatch/2021-04-13.dat を追加。
    有効にするには、raspirecの再起動とEPG更新が必要です。

ver 1.0.2         2021/04/04
  ・不具合修正
    *「録画済み一覧」から自動予約を削除した時に、「録画済み一覧」
       の画面に戻らない件を修正

ver 1.0.1         2021/04/02
  ・機能追加
    * top画面に、「過去3日間の録画エラー数」の表示を追加

ver 1.0.0         2021/03/20
  ・機能追加
    * tspacketchk によるパケットの健全性チェックの組み込み
      設定方法は、
      https://kaikoma-soft.github.io/src/raspirec-packetchk.html を参照

  ・機能追加が一段落したのでβ版からリリース版に昇格

ver 0.6.4         2021/02/20
  ・稀に Ubuntu 20 で waitpid()の競合が起き「終了処理未了」になる件の対策

ver 0.6.3         2021/02/09
  ・BS再編の記録を残すように。
  ・改良
    *「その他」->「コントロールパネル」->「チャンネル情報」で、削除ボ
      タンを常に押せるように変更。

ver 0.6.2         2021/01/05
  ・改良
    * ファイル転送の2重実行をしないように排他制御を導入
  ・不具合修正
    * 番組表等でタイトルに " が含まれると表示が崩れる件を修正

ver 0.6.1         2020/11/14
  ・機能追加
    * 自動予約一覧とフィルター一覧で、共通だった表のソートの設定値の保
      存を分離した。

ver 0.6.0         2020/10/16

  ・機能追加
    * 自動予約一覧で、表にジャンルの項を追加
    * 自動予約一覧で、表のソート機能を追加
    * 番組検索の画面に削除ボタンを追加(ただし新規の場合は出さない)
    * 番組検索の画面に検索文字列をコピーするボタンを追加
    * --config オプションの追加
    * 番組表や検索結果から自動予約する時の初期値でジャンルの小項目は
      "全て" に置き換える

  ・不具合修正
    * 番組表->詳細->録画予約 した時のダイアログの時短の文章の文字色が
      灰色になっているのを修正
    * 局ごとの番組表から詳細ダイアログで「視聴」ボタンが押せてしまう件
      を修正

ver 0.5.0         2020/06/17

  ・機能追加
    * mpv を使ったモニタ機能の追加
      設定方法は、https://kaikoma-soft.github.io/raspirec-option.html を参照
    * プルダウンの config と help をコントロールパネル中に移動 and 名称変更

ver 0.4.1         2020/06/01

  ・機能追加
    * EPG取得の禁止時間帯を設定出来るように config.rb の EpgBanTime
      を追加
      ( https://github.com/kaikoma-soft/raspirec/issues/7 を参照)

ver 0.4.0         2020/05/03

  ・機能追加
    * EPGデータにパッチを当てる機能を追加
      2020/03/末の FOXスポーツエンタ停波による影響で、EPGデータと実際
      のslot番号がズレる現象があり、放送大学とＢＳスカパー！が受信出来
      ない問題をこれで解消する。
    * チャンネル毎に更新日時を記録するように。
    * EPGチャンネルデータに変更があった場合に、DB をアップデートするように。
    * コントロールパネル中に「チャンネル情報」を追加
    * チャンネル情報を一覧で表示し、不要になったチャンネルを削除出来る
      ように。

ver 0.3.5         2020/04/16

  ・不具合修正
    * フイルターの検索結果表示で、録画中のものが×と誤表示されてしまう
      件の修正

ver 0.3.4         2020/04/12

  ・不具合修正
    * フイルターの検索結果表示で、過去になった番組も表示してしまう件の
      修正

ver 0.3.3         2020/03/29

  ・「自動予約」ボタンを押して番組検索に遷移した時に、題名および検索文
     字列を番組タイトルから生成する為の文字削除をユーザーがカスタマイ
     ズ出来るように config.rb に項目追加

ver 0.3.2         2020/03/18

  ・不具合修正
    * 自動予約で放送時間に条件を付けていた場合、スケジューラーが時短
      を行わないバグの修正

ver 0.3.1         2020/02/10

  ・番組検索に放送時間の項目を追加
  ・局毎番組表にも「オプション」のボタンを追加。


ver 0.3.0         2020/02/01

  ・番組表に「オプション」のボタンを追加。
    そのダイアログで、番組表の縦軸、横軸等を可変に出来る機能を追加
  ・コード整理：ajax の書き方の統一
  ・不具合修正
    * 予約一覧のダイアログで、一部の文字が薄かったのを修正

ver 0.2.3         2019/12/18

  ・出力するTSファイルのファイル名をカスタマイズ出来るように,
    config.rb に TSnameFormat を追加。
    ドキュメントも追加したので、詳細は doc/TS_file_name.md を参照。

  ・不具合修正
    * 録画中止の後にそれと重なる予約をすると、予約状況表にズレが生じる
      不具合の修正

ver 0.2.2         2019/12/16
  ・不具合修正
    * デバック用の require 'sinatra/reloader' が残っていたのを削除
  ・Readme.md の詳細を github pages に移動

ver 0.2.1         2019/11/07
  ・モバイル版で、メニューのモニタ機能のOn/Off の制御が抜けていたのを修正

ver 0.2.0         2019/11/01
  ・機能追加
    * 番組検索の局名から局番組表へ飛べるように
    * DISK削除で、空のディレクトリも削除するように。
    * html5 video モニタ機能の追加
  ・不具合修正
    * 録画済み一覧で、既に削除済みの自動登録へのリンクをクリックすると
      帰ってこない件を修正

ver 0.1.2         2019/10/01
  ・EPG 取得時 Timeout を伸ばす。

ver 0.1.1         2019/09/27
  ・EPG 取得時の fork に Timeout処理を追加
  ・データ放送等を番組表から除外する機能追加
  ・ディスク容量確保動作を１日一回から 6時間毎に変更
  ・config.rb に StationPage (番組表で１ページ当たりの放送局数)を追加
  ・予約削除する時の確認ダイアログに、対象のタイトルを表示するように
  ・Bug Fix
     * 予約した後に番組名の変更があった場合に追従しなかったのを修正



ver 0.1.0         2019/09/23
  ・β版 公開初版


src/views/config.rb
#!/usr/local/bin/ruby 
# -*- coding: utf-8 -*-

#
#  コントロール パネル
#

class Config

  Text = {
    "httpd" => <<-EOS,
Http_port    httpd のポート番号を指定する。<BR>
+            例： http://localhost:4567/
EOS
    "dirFile" => <<-EOS,
Recpt1_cmd   recpt1/recdvb コマンドの path を指定する。
Recpt1_opt   recpt1/recdvb コマンドのオプションを指定する。
+            b25 デコードをする場合は --b25 を指定する。
Epgdump      epgdump コマンドの path を指定する。
BaseDir      raspirec がインストールされているディレクトリを設定する。
+            ( raspirec.rb があるディレクトリ )
DataDir      データベースや録画したファイルの置き場所を指定する。
EOS
    "recTime" => <<-EOS,
Start_margin  番組開始前の録画マージンを指定(秒)。
+             録画は PC の時計を基準に、開始します。
After_margin  番組終了後の録画マージンを指定(秒)。
+             録画は PC の時計を基準に、終了します。
Gap_time      録画が連続し、前番組が時短になった場合の、
+             前番組録画終了から次番組開始間隔を指定(秒)。<br>
+             前番組は Start_margin +  Gap_time の秒数だけ録画時間が削られます。
+              <br>TVチューナー依存ですが、あまり短いと動作が不安定になります。
EOS
    "tuner" => <<-EOS,
GR_tuner_num         地デジ チューナー数
BSCS_tuner_num       BS/CS チューナー数
GBC_tuner_num        ３波(地デジ/BS/CS)チューナー数
Total_tuner_limit    同時に動作するチューナー数に制限を掛ける場合に指定する。
+                    使用しない場合は false 
EOS

    "epg" => <<-EOS,
GR_EPG_channel      地デジ EPG 受信局を物理チャンネルで指定する。<br>
+                   東京スカイツリーならば 
+                   %w( 27 26 25 24 22 23 21 16 ) になります。
+       <a href="https://www.maspro.co.jp/contact/channel.pdf#page=4"> 参考 </a>
BS_EPG_channel      BS EPG 受信局を指定する。BSを受信しない場合は空にする。
+                   <br>通常の番組情報は、どれか 1局だけ受信すれば十分だが、
+                   詳細情報を取得するには、その情報を取りたい局を指定する必要がある。
CS_EPG_channel      CS EPG 受信局を指定する。CSを受信しない場合は空にする。
+                   指定はトランスポンダ(CS2〜CS24)単位で行う。
GR_EpgRsvTime       地デジ EPG受信時間 (秒)
BS_EpgRsvTime       BS EPG受信時間 (秒)
CS_EpgRsvTime       CS EPG受信時間 (秒)
EPGperiod           EPG 取得周期 (H)
EpgBanTime          EPG の取得禁止時間帯の指定。(24H制 時間単位) 
+                   1時から5時までを禁止したい場合は、[ 1,2,3,4,5 ] と指定する。
+                   使用しない場合は nil
EPG_tuner_limit     EPG取得時の使用チュナー数に制限を掛ける場合に指定する。使用しない場合は false
EOS

    "RecExt" => <<-EOS,
AutoRecExt          自動録画延長機能  true で有効。 詳細は
+                   <a href="https://kaikoma-soft.github.io/raspirec-RecExt.html"> こちら </a> を参照
ARE_sampling_time   番組終了の n 秒前に EPG 採取
ARE_epgdump_opt     最後尾切り出しの epgdump のオプション
EOS
    
    "dlgOpt" => <<-EOS,
D_FreeOnly          無料放送のみ
D_dedupe            重複予約は無効化する
D_jitan             チューナー不足の場合に時短を許可
EOS

    "tsft" => <<-EOS,
TSFT          true=転送機能有効 true以外=無効<br>
+             この機能を使うには、送り先のホストに対して,パスワードなしで 
+              ssh,scpアクセス可能なように設定されていることが必要です。
TSFT_host     送り先 ホスト名
TSFT_user     送り先 login名
TSFT_toDir    送り先Dir
TSFT_rate     想定転送速度 ( Mbyte/秒 )<br> 
+             この数字を使って、空き時間に転送する／しないを判断します。
EOS

    "hlsMon" => <<-EOS,
MonitorFunc       true=モニタ機能を有効, false=無効。 無効の場合は、下記のパラメータは無視される。
+                設定方法は <a href="https://kaikoma-soft.github.io/raspirec-option.html"> こちら </a> を参照して下さい。
StreamDir        ストリーム出力先ディレクトリ
MonitorWidth     モニタ画面の横幅
HlsConvCmd       HLS変換コマンド
#
#
EOS

    "mpvMon" => <<-EOS,
MPMonitor        true = mpv モニタ機能を有効、false = 無効 <br>無効の場合は、下記のパラメータは無視される。
+                設定方法は <a href="https://kaikoma-soft.github.io/raspirec-option.html"> こちら </a> を参照して下さい。
DevAutoDetection  DeviceList_GR, DeviceList_BSCS, DeviceList_GBC を自動設定する。 
+                 <br> true = 有効。 
+                 詳細は <a href="../help#DevAutoDetection"> 補足説明 </a> を参照して下さい。
DeviceList_GR    地デジ チューナーのデバイスファイルを指定する。
DeviceList_BSCS  BS,CS チューナーのデバイスファイルを指定する。
DeviceList_GBC   三波共用(地デジ/BS/CS)チューナーのデバイスファイルを指定する。
Mpv_cmd          mpv の実行ファイル名(絶対パス)
Mpv_opt          mpv の引数を指定する。
RemoteMonitor    表示するマシンが別の場合 ture, 同一の場合に false
UDPbasePort      使用する UDP のポート番号、デバイスの数だけプラスされる。
XServerName      RemoteMonitor が true の場合に、
+                mpvを実行、表示するマシンのマシン名を設定する。
RecHostName      RemoteMonitor が true の場合に、
+                チューナー(raspirecが実行されている)のマシン名を設定する。
Lsof_cmd         lsof コマンドへのパス。Ubuntu では /usr/bin/lsof 
Browser_cmd      番組表を表示するブラウザの実行ファイル名(絶対パス)
RaspirecTV_font  フォント指定
RaspirecTV_GEO   座標指定 ( WxH+X+Y or X+Y )
RaspirecTV_SOCAT socat の実行ファイル名(絶対パス)(RemoteMonitor が true の時のみ)
EOS

    "PacketChk" => <<-EOS,
PacketChk_enable    パケットチェックを有効にするか。true で有効
PacketChk_cmd       tspacketchk のコマンドパスを指定する
PacketChk_opt       tspacketchk のオプションを指定する
PacketChk_threshold エラーと判定する閾値(画面上で赤枠をつける)
PacketChk_rate      空き時間内に終わるか計算する為の想定速度 ( Mbyte/秒 )
EOS

    "other" => <<-EOS,
TSnameFormat      TSファイル名の生成ルール。詳細は
+                 <a href="/help#TSnameFormat"> 補足説明 </a> を参照して下さい。
LogSaveDay        ログの保持期間(日)
RsvHisSaveDay     録画済み記録の保持期間(日)
DiskKeepPercent   録画したTSファイルを古い順に削除して指定したDisk容量(%)を
+                 確保する。(指定するのは空き容量)<br> 
+                 機能を無効にする場合は false を指定する。
Local_jquery      オフライン環境で動作させる為に jquery, materialize の
+                 ライブラリをローカルにコピーした場合に、true にする。<br> 
+                 詳細は doc/jquery_local.md を参照の事。 通常は false 
StationPage       番組表で、1ページ当たりの放送局数 (個) の初期値
Debug             ture で Debug モード。ログファイルを出力するようになる。
+                 オプションで、 -d を指定するのと同じ。
Debug_mem         ture で メモリの消費量をモニタするようになる。
TitleRegex        「自動予約」ボタンを押して番組検索に遷移した時に、
+                 番組タイトルから「題名」を生成する為の、余計な文字を削除する
+                 正規表現の配列
SearchStringRegex  「自動予約」ボタンを押して番組検索に遷移した時に、
+                 番組タイトルから「検索文字列」を生成する為の、余計な文字を
+                 削除する正規表現の配列
EpgPatchEnable    EPGPatch機能を有効にする。 false = 無効。デフォルトは有効<br>
+                 dvbドライバー + recdvb の場合は false に設定する。
EOS
  }

  Title = {
    "init"    => "はじめに",
    "httpd"   => "httpd",
    "dirFile" => "ディレクトリ、ファイル関係",
    "recTime" => "録画タイミング関係",
    "tuner"   => "チューナー関係",
    "epg"     => "EPG関係",
    "RecExt"  => "録画自動延長",
    "dlgOpt"  => "ダイアログのオプション初期値",
    "tsft"    => "TSファイル転送",
    "hlsMon"  => "HLS モニタ機能",
    "mpvMon"  => "mpv モニタ機能",
    "other"   => "その他",
  }
   
  def initialize(  )
    @data = {}
    @list = {}
    pkey = nil
    Text.each_pair do |ref,v|
      @list[ ref ] ||= []
      v.each_line do |line|
        line.chomp!
        if line =~ /^#/
          next
        elsif line =~ /^\+(.*)/
          @data[ pkey ] += $1.strip if pkey != nil
        elsif line =~ /(\w+)\s+(.*)/
          key, val = $1, $2
          @data[ key ] = val
          @list[ ref ] << key
          pkey = key
        elsif line == ""
          next
        else
          pp line
          raise
        end
      end
    end
  end

  #
  #  
  #
  def print( key )
    text = @data[ key ].strip.gsub(/\n/,'<br>')
    if Object.const_defined?( key ) == true
      val  = Object.const_get( key )
    else
      val = "未定義"
    end
    val = "nil" if val == nil
    ret = <<-EOS
    <tr>
       <td> #{key} </td>
       <td> #{val} </td>
       <td> #{text}</td>
    </tr>
    EOS
    ret
  end

  #
  #  
  #
  def header(  )
    ret = <<-EOS
    <tr>
       <th> 定数名 </th>
       <th> 現在値 </th>
       <th> 説明   </th>
    </tr>
    EOS
    ret
  end

  def printMarkD()
    @list.each_pair do |k,v|
      title = Title[ k ]
      printf("\n\n\#\#\#\# %s\n", title )
      puts( <<EOS )

| 定数名             | 説明 |
|--------------------|------|
EOS
      v.each do |k2|
        text = @data[ k2 ]
        #text.gsub(/(\$\{.*?\})/,'\\1')
        printf("|%s|%s|\n", k2, text )
      end
    end
  end

  def printHTML()
    ret = []
    @list.each_pair do |k,v|
      title = Title[ k ]
      ret << sprintf(%Q(<li class="title">\n\t<a name="%s">%s</a>\n),k,title )
      ret.push( <<EOS )
          <table class="striped">
            <tr>
               <th> 定数名 </th>
               <th> 現在値 </th>
               <th> 説明   </th>
            </tr>
EOS
      v.each do |k2|
        text = @data[ k2 ]
        if Object.const_defined?( k2 ) == true
          val  = Object.const_get( k2 )
        else
          val = "未定義"
        end
        val = "nil" if val == nil
        ret << "<tr>"
        ret << sprintf("\t<td>%s</td><td>%s</td><td>%s</td>\n", k2, val,text )
        ret << "</tr>"
      end
      ret.push( <<EOS )
   </table>
  </li>
EOS
    end
    ret.join("\n")
  end

end


if $0 == __FILE__
  #
  #  markdown 形式で出力する。
  #
  Config.new.printMarkD()
  
end
